In [10]:
import os
import sys
import re
import pandas as pd

# Import the text file
text = open("magic.txt","r").read()

### Task 1: Store the Card Details in Well Structured format


A class `MagicCard` contains the objects { CardName, CardCost, CardType, CardEffect }

Also to view the card detail a to_dict method is created


In [11]:
# Store Card Data in Structured Format
class MagicCard:
    def __init__(self, name, cost, card_type, effect):
        self.CardName = name
        self.CardCost = cost
        self.CardType = card_type
        self.CardEffect = effect

    def to_dict(self):
        return {
            'CardName': self.CardName,
            'CardCost': self.CardCost,
            'CardType': self.CardType,
            'CardEffect': self.CardEffect
        }

### Task 2 & 3 Make a list of lines for each Card Transform the textfile to a dataframe

The text file is split into list of lines representing Card and its details.

The information is then added in structured class and also converted to a dataframe

In [12]:
lines = text.split("\n")
cards: list[MagicCard] = []
def parse_card_data(lines):
    for line in lines:
        if not line.strip():
            continue
        
        card_pattern = re.compile(r"^CardName:\s*(.*?)\s*CardCost:\s*(.*?)\s*CardType:\s*(.*?)\s*CardEffect:\s*(.*)$")
        match = card_pattern.match(line)
        # print(match)
        if match:
            CardName= match.group(1)
            CardCost= match.group(2)
            CardType= match.group(3)
            CardEffect= match.group(4)
            cards.append(MagicCard(CardName, CardCost, CardType, CardEffect))

parse_card_data(lines)
            

In [13]:
print(cards[0].to_dict())

{'CardName': 'Absorb', 'CardCost': '{W}{U}{U}', 'CardType': 'Instant', 'CardEffect': 'Counter target spell. You gain 3 life.'}


In [14]:
#Export cards to pandas DataFrame

card_dicts = [card.to_dict() for card in cards]

# Create the pandas DataFrame (df)
df = pd.DataFrame(card_dicts)

# Rename columns as requested ("Name", "Cost", "Type", "Effect")
df = df.rename(columns={
    'CardName': 'Name',
    'CardCost': 'Cost',
    'CardType': 'Type',
    'CardEffect': 'Effect'
})



In [15]:
len(df)

1419

In [16]:
print(df.head())

                  Name       Cost                            Type  \
0               Absorb  {W}{U}{U}                         Instant   
1  Acclaimed Contender     {2}{W}       Creature — Human Knight   
2       Act of Treason     {2}{R}                         Sorcery   
3       Aerial Assault     {2}{W}                         Sorcery   
4         Aeromunculus  {1}{G}{U}  Creature — Homunculus Mutant   

                                              Effect  
0             Counter target spell. You gain 3 life.  
1  When Acclaimed Contender enters the battlefiel...  
2  Gain control of target creature until end of t...  
3  Destroy target tapped creature. You gain 1 lif...  
4  Flying {2}{G}{U}: Adapt 1. (If this creature h...  


### Task 4: Find which type of Card appears the most
Which of the words “Creature”, “Sorcery”, “Instant”, “Enchantment” and “Artifact” appears
most often in all the texts within the “Type” column?

In [17]:
findWords =["Creature", "Sorcery", "Instant", "Enchantment", "Artifact"]
count_dict = {}
for word in findWords:
    for index, row in df.iterrows():
        if word in row['Type']:
            if word in count_dict:
                count_dict[word] += 1
            else:
                count_dict[word] = 1
print(count_dict)

{'Creature': 733, 'Sorcery': 185, 'Instant': 215, 'Enchantment': 85, 'Artifact': 108}


### Answer to task 4
There are in total of 1419 Cards
| **Card Type** | **Count** |
| :--- | :--- |
| Creature | 733 |
| Sorcery | 185 |
| Instant | 215 |
| Enchantment | 85 |
| Artifact | 108 |

In [18]:
# Check answer manually in CSV using search
df.to_csv('card_data.csv', index=False)